mlflow/sklearn/__init__.py

```diff
@@ -1421,65 +1424,129 @@ def patched_model_score(original, self, *args, **kwargs):
+
+    def patch_estimator_method_if_available(class_def, func_name, patched_fn, manage_run):
+        if not hasattr(class_def, func_name):
+            return
+
+        original = getattr(class_def, func_name)
+        if isinstance(original, property):
+            # Check whether we should patch the estimator method, will check:
+            # * the estimator has the "func_name" attribute
+            # * the attribute is not a property, this is for filtering out cases like
+            #   `LocalOutlierFactor.predict` which is a property
+            #   A couple of estimators use property methods to return fitting functions,
+            #   rather than defining the fitting functions on the estimator class directly.
+            #
+            #   Example: https://github.com/scikit-learn/scikit-learn/blob/0.23.2/sklearn/neighbors/_lof.py#L183  # noqa
+            #
+            #   We currently exclude these property fitting/inference methods from patching because
+            #   it's challenging to patch them correctly.
+            #
+            #   Excluded fitting/inference methods:
+            #   - sklearn.cluster._agglomerative.FeatureAgglomeration.fit_predict
+            #   - sklearn.neighbors._lof.LocalOutlierFactor.fit_predict
+            #   - sklearn.neighbors._lof.LocalOutlierFactor.predict
+            #   - sklearn.multioutput.MultiOutputClassifier.predict_proba
+            #   - sklearn.svm._classes.NuSVC.predict_proba
+            #   - sklearn.linear_model._stochastic_gradient.SGDClassifier.predict_proba
+            #   - sklearn.svm._classes.SVC.predict_proba
+            #   - sklearn.ensemble._voting.VotingClassifier.predict_proba
+            #   - sklearn.svm._classes.NuSVC.predict_log_proba
+            #   - sklearn.linear_model._stochastic_gradient.SGDClassifier.predict_log_proba
+            #   - sklearn.svm._classes.SVC.predict_log_proba
+            #
+            #   You can list property fitting methods by inserting "print(class_def, func_name)"
+            #   in the if clause below.
+            #
+            #   TODO: Convert these property methods into normal methods decorated by `@available_if`
+            #     See https://github.com/scikit-learn/scikit-learn/issues/20630
+            return
+
+        # Retrieve raw attribute while bypassing the descriptor protocol
+        raw_original_obj = gorilla.get_attribute(class_def, func_name)
+
+        if original is not raw_original_obj and \
+                not hasattr(raw_original_obj, 'delegate_names') and \
+                not hasattr(raw_original_obj, 'check'):
+            # Unknown getter found. Cannot patch.
+            return
+
+        safe_patch(
+            FLAVOR_NAME, class_def, func_name, patched_fn, manage_run=manage_run,
+        )
+
+        if original is not raw_original_obj:
+            safe_patch_fn = getattr(class_def, func_name)
+
+            if hasattr(raw_original_obj, 'delegate_names'):
```

Comment:
Can we use `isinstance(raw_original_obj, _IffHasAttrDescriptor)` here

Reply:
I doubt they may refactor the code and move `_IffHasAttrDescriptor` to other modules or change class name. seems checking attribute is more stable.

Reply:
got it.